distribution/pull_v2.go

```diff
@@ -501,6 +519,29 @@ func (p *v2Puller) pullSchema2(ctx context.Context, ref reference.Named, mfst *s
 		return imageID, manifestDigest, nil
 	}
 
+	var descriptors []xfer.DownloadDescriptor
+
+	// Note that the order of this loop is in the direction of bottom-most
+	// to top-most, so that the downloads slice gets ordered correctly.
+	for _, d := range mfst.Layers {
+		layerDescriptor := &v2LayerDescriptor{
+			digest:            d.Digest,
+			repo:              p.repo,
+			repoInfo:          p.repoInfo,
+			V2MetadataService: p.V2MetadataService,
+		}
+
+		if d.MediaType == schema2.MediaTypeForeignLayer && len(d.URLs) > 0 {
+			if runtime.GOOS != "windows" {
```

Comment:
Rather than the push/pull code looking at OS directly, I'd suggest having a function in the `layer` package like `ForeignSourceSupported`.

Another way to do it that could be interesting is to remove `ForeignSource` from the `Layer` interface, and only define the method on Windows (by putting it in a `*_windows.go` file). Then we could put it in its own interface definition:

```go
type ForeignSourcer interface {
	// ForeignSource returns the foreign reference information for
	// the layer.
	ForeignSource() *distribution.Descriptor
}
```

Then the download manager could do something like:

```go
_, foreignSrcSupported := ldm.layerStore.(layer.ForeignSourcer)
if descriptor.foreignSrc != nil && !foreignSourceSupported {
        return errors.New("foreign layers are not supported on this OS")
}
```

The advantage of doing it this way is that we wouldn't pollute the `Layer` interface, so other OS' never have to implement it or deal with it, and things like mocks in unit tests don't need to implement the `ForeignSource` method.